File: mobile/targets/watch/Push.swift
// Registering this watch for push.
//
// The three tip-off notices do not come through here and should not: those
// are scheduled locally from cached fixtures (Notifications.swift), so they
// fire in a gym with no signal. Push exists for what the watch cannot work
// out on its own — a captain approving your sub, a teammate flagging out —
// which previously waited on a background refresh watchOS granted whenever
// it felt like it.
//
// watchOS 6 and later lets a watch app hold its **own** APNs token, which is
// the only reason this is possible at all: the phone is in a drawer at home,
// so a token belonging to the phone app would be no use.
//
// The token is registered against the signed-in account, and retired on sign
// out. Registration is idempotent and cheap, so it runs on every launch
// rather than trying to remember whether it has run before — Apple can
// reissue a token after a restore, and a device quietly holding a stale one
// is exactly the bug that is impossible to notice.

import Foundation
import UserNotifications
import WatchKit

@MainActor
enum Push {
  /// Set by the delegate the moment Apple hands one over.
  private(set) static var deviceToken: String?

  /// The API this posts to. Not Supabase directly: the token has to be tied
  /// to the caller's identity, and letting the app write its own row means
  /// trusting the client with whose account it belongs to. The route does it
  /// under RLS instead.
  private static var registerURL: URL? {
    URL(string: "https://www.intramural.app/api/push/register")
  }

  /// Ask once, then tell watchOS to go get a token. Declining is fine —
  /// nothing here is the only route to any piece of information.
  static func start() {
    Task {
      let granted = (try? await UNUserNotificationCenter.current()
        .requestAuthorization(options: [.alert, .sound])) ?? false
      guard granted else { return }
      WKApplication.shared().registerForRemoteNotifications()
    }
  }

  /// Called from the delegate with Apple's raw token.
  static func adopt(_ data: Data) {
    deviceToken = data.map { String(format: "%02x", $0) }.joined()
  }

  /// Hand the token to the server, tied to whoever is signed in.
  ///
  /// Silent on failure by design: a watch that cannot register is a watch
  /// that gets its notices a little later, not a watch that should show an
  /// error over the one screen the student actually needs.
  static func register(api: Api) async {
    guard let token = deviceToken, let url = registerURL else { return }
    guard let accessToken = try? await api.validToken() else { return }

    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
    req.httpBody = try? JSONSerialization.data(withJSONObject: [
      "token": token,
      "platform": "watchos",
      // The APNs topic is the *watch app's* bundle id, which is not the
      // phone's — read it rather than hard-coding, so a rename cannot leave
      // this silently pointing at an app that no longer exists.
      "bundleId": Bundle.main.bundleIdentifier ?? "",
    ])
    _ = try? await URLSession.shared.data(for: req)
  }

  /// Retire this token when the account signs out.
  ///
  /// Takes the bearer rather than the client because the caller has to
  /// capture it *before* signing out — an async refresh would be racing the
  /// session it depends on.
  ///
  /// Best-effort, and safe to lose: registration upserts on the device token
  /// itself, so the next person to sign in on this watch takes ownership of
  /// the row regardless. That upsert is the actual guarantee that nobody
  /// inherits someone else's notifications; this is just tidying up.
  static func unregister(bearer accessToken: String) async {
    guard
      let token = deviceToken,
      let base = registerURL,
      var parts = URLComponents(url: base, resolvingAgainstBaseURL: false)
    else { return }
    parts.queryItems = [URLQueryItem(name: "token", value: token)]
    guard let url = parts.url else { return }

    var req = URLRequest(url: url)
    req.httpMethod = "DELETE"
    req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
    _ = try? await URLSession.shared.data(for: req)
  }
}

File: mobile/lib/push.ts
/**
 * Registering this phone for push.
 *
 * The token is the *raw APNs device token*, not an Expo push token — the
 * server sends through Apple directly (lib/notify/apns.ts on the web side),
 * so Expo's relay never enters the picture. `getDevicePushTokenAsync` hands
 * back exactly what Apple issued.
 *
 * Registration runs on every signed-in launch rather than once: Apple can
 * reissue the token after a restore, registration upserts on the token
 * itself, and a device quietly holding a stale registration is the failure
 * nobody ever notices. Silent on every error by design — a phone that can't
 * register is a phone that gets its announcements a little later, not an
 * error worth a dialog.
 */

import * as Notifications from "expo-notifications";
import Constants from "expo-constants";
import { Platform } from "react-native";

const SITE =
  process.env.EXPO_PUBLIC_SITE_URL ?? "https://www.intramural.app";

/** Show announcements even while the app is foregrounded — a commissioner's
    "gym closed" must not depend on the app being closed to be seen. */
export function configureForegroundPresentation() {
  Notifications.setNotificationHandler({
    handleNotification: async () => ({
      shouldShowBanner: true,
      shouldShowList: true,
      shouldPlaySound: false,
      shouldSetBadge: false,
    }),
  });
}

export async function registerForPush(accessToken: string): Promise<void> {
  // Simulators have no APNs; Android would need a different pipeline.
  if (Platform.OS !== "ios") return;
  try {
    const perm = await Notifications.requestPermissionsAsync();
    if (!perm.granted && !perm.ios?.status) return;

    const device = await Notifications.getDevicePushTokenAsync();
    const token = String(device.data);
    // Apple's token is 32 bytes hex; anything else here is a simulator or
    // an Expo Go session, neither of which can receive APNs anyway.
    if (!/^[0-9a-fA-F]{64}$/.test(token)) return;

    await fetch(`${SITE}/api/push/register`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${accessToken}`,
      },
      body: JSON.stringify({
        token,
        platform: "ios",
        bundleId:
          Constants.expoConfig?.ios?.bundleIdentifier ?? "app.intramural.ios",
      }),
    });
  } catch {
    // Declined permission, no network, misconfigured build — all fine.
  }
}

/** Retire this phone's token on sign-out so the next account on this device
    does not inherit the last one's announcements. */
export async function unregisterForPush(accessToken: string): Promise<void> {
  if (Platform.OS !== "ios") return;
  try {
    const device = await Notifications.getDevicePushTokenAsync();
    const token = String(device.data);
    if (!/^[0-9a-fA-F]{64}$/.test(token)) return;
    await fetch(
      `${SITE}/api/push/register?token=${encodeURIComponent(token)}`,
      {
        method: "DELETE",
        headers: { Authorization: `Bearer ${accessToken}` },
      },
    );
  } catch {
    // Best-effort: registration upserts on the token, so the next sign-in
    // takes ownership of the row regardless.
  }
}
